### PR TITLE
feat(oauth2): new library to create oauth2 tokens

### LINKED ISSUE #11920: Create a public API to get access token
The APIs to get (and cache) access tokens are all in `*internal*` namespaces.  We need a simple API to create access tokens.

Maybe something like this:

```cc
namespace google::cloud {
struct AccessToken { std::string token; std::chrono::time_point expiration; };

class AccessTokenGenerator {
 public:
   virtual ~AccessTokenGenerator() = default;
   virtual StatusOr<AccessToken> GetToken() = 0;
};

std::shared_ptr<AccessTokenGenerator> MakeAccessTokenGenerator(std::shared_ptr<Credentials>);
```

That is a fairly minimal API, whatever our implementation details we will always have enough to support it. 

We need to think about the name of the libraries too.  All of this lives in `rest_internal` which suggests it is not a public library.  We may need to create `google_cloud_cpp_auth` ?

### PR DESCRIPTION
**Pull request by coryan**
This is basically a wrapper around some classes in the `rest_internal` library.  We need a new library because applications should not link `rest_internal`: all libraries with `internal` in their name are not part of the public API.

Fixes #11920

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12064)

### CODE REVIEW

**Review by scotthart — COMMENTED** [review]
If we anticipate more public OAuth2 files/classes, should we consider adding a google/cloud/oauth2 directory for these to live in, as opposed to google/cloud?

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12064#-:-N_4uGexBZVm1eoZ3Yvb:bcob5ws)* status: 0 of 14 files reviewed, all discussions resolved

File: ci/cloudbuild/builds/check-api.sh [hunk 1]
```diff
@@ -39,14 +39,14 @@ io::run cmake "${cmake_args[@]}" \
   -DGOOGLE_CLOUD_CPP_ENABLE="${ENABLED_FEATURES}" \
   -DCMAKE_CXX_FLAGS="-Og -Wno-maybe-uninitialized"
 io::run cmake --build cmake-out
-io::run cmake --install cmake-out >/dev/null
+io::run cmake --install cmake-out #>/dev/null
```
Comment (devbww) [inline ci/cloudbuild/builds/check-api.sh:42]: Remnant or feature?

File: google/cloud/access_token_generator.cc [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+      std::shared_ptr<google::cloud::oauth2_internal::Credentials> impl)
+      : impl_(std::move(impl)) {}
+  ~AccessTokenGeneratorImpl() override = default;
+
+  StatusOr<AccessToken> GetToken() override {
+    return impl_->GetToken(std::chrono::system_clock::now());
+  }
+
+ private:
+  std::shared_ptr<google::cloud::oauth2_internal::Credentials> impl_;
```
Comment (devbww) [inline google/cloud/access_token_generator.cc:35]: Is `impl_` the right name, versus, say `creds_`?

File: google/cloud/access_token_generator.cc [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+};
+
+}  // namespace
+
+std::shared_ptr<AccessTokenGenerator> MakeAccessTokenGenerator(
+    std::shared_ptr<Credentials> credentials) {  // NOLINT
```
Comment (devbww) [inline google/cloud/access_token_generator.cc:41]: Aside: What lint is that?

File: ci/cloudbuild/builds/check-api.sh (hunk 1 above)
Comment (coryan) [reply ci/cloudbuild/builds/check-api.sh]: Remnant, fixed.

File: google/cloud/access_token_generator.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/access_token_generator.cc]: I can argue both sides of that.  And sound like I believe both of them too!  Changed to `creds_`.

File: google/cloud/access_token_generator.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/access_token_generator.cc]: Passing by value and only using as a `const&`.  I think we should consume `Credentials const&`, but that requires further refactoring.  I will do that and then rebase this PR.

**Review by coryan — COMMENTED** [review]
I am thinking about it.  Sounds like a good idea. Let me do some other refactoring first and maybe discuss in the meeting tomorrow (won't merge until after that discussion).

> Reviewed 11 of 13 files at r1, 1 of 1 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12064#-:-N_6BE809yuvuO-Mzk8w:b-1bhq4u)* status: all files reviewed, 3 unresolved discussions (waiting on @devbww)

**Review by devbww — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12064)* status: all files reviewed, 1 unresolved discussion (waiting on @coryan)

**Review by devbww — APPROVED** [review]

**Pull request merged**